server/lib/nl/detection/context.py

```diff
@@ -0,0 +1,272 @@
+
+  if is_explore:
+    max_returned_vars = _MAX_RETURNED_VARS_EXPLORE
+  else:
+    max_returned_vars = _MAX_RETURNED_VARS_CHAT
+
+  # 6. Populate the returned dict
+  data_dict.update({
+      Params.ENTITIES.value: places,
+      Params.VARS.value: vars[:max_returned_vars],
+      Params.SESSION_ID.value: uttr.session_id,
+      Params.CMP_ENTITIES.value: cmp_places,
+      Params.CMP_VARS.value: cmp_vars[:max_returned_vars],
+      Params.CHILD_TYPE.value: '' if not place_type else place_type.value,
+  })
+
+  # 6. Set the detected params in uttr ctx and clear past contexts.
+  uttr.insight_ctx = data_dict
+
+
+def _detect_vars(uttr: nl_uttr.Utterance, is_cmp: bool,
+                 hoist_topics: bool) -> List[str]:
+  svs = []
+  cmp_svs = []
+  if is_cmp:
+    # Comparison
+    if dutils.is_multi_sv(uttr.detection):
+      # Already multi-sv, nothing to do in `uttr`
+      svs, cmp_svs = _get_multi_sv_pair(uttr, hoist_topics)
+      uttr.sv_source = nl_uttr.FulfillmentResult.CURRENT_QUERY
+    else:
+      if uttr.svs and uttr.prev_utterance and uttr.prev_utterance.svs:
+        # Set `multi-sv`
+        svs = uttr.svs
+        cmp_svs = uttr.prev_utterance.svs
+        # Set a very basic score. Since this is only used by correlation
+        # which will do so regardless of the score.
+        uttr.multi_svs = dutils.get_multi_sv(svs, cmp_svs, 0.51)
+        # Important to set in detection since `correlation.py` refers to that.
+        uttr.detection.svs_detected.multi_sv = uttr.multi_svs
+        uttr.sv_source = nl_uttr.FulfillmentResult.PARTIAL_PAST_QUERY
+  else:
+    # No comparison.
+    if uttr.svs:
+      svs = uttr.svs
+      uttr.sv_source = nl_uttr.FulfillmentResult.CURRENT_QUERY
+    else:
+      # Try to get svs from context.
+      if uttr.prev_utterance and uttr.prev_utterance.svs:
+        svs = uttr.prev_utterance.svs
+        uttr.svs = svs
+        uttr.counters.info('insight_var_ctx', svs)
+        uttr.sv_source = nl_uttr.FulfillmentResult.PAST_QUERY
+
+  return svs, cmp_svs
+
+
+def _get_comparison_or_correlation(
+    uttr: nl_uttr.Utterance) -> ClassificationType:
+  for cl in uttr.classifications:
+    if cl.type in [
+        ClassificationType.COMPARISON, ClassificationType.CORRELATION
+    ]:
+      return cl.type
+  # Mimic NL behavior when there are multiple places.
+  if len(uttr.places) > 1:
+    return ClassificationType.COMPARISON
+  return None
+
+
+def _get_multi_sv_pair(uttr: nl_uttr.Utterance,
+                       hoist_topics: bool) -> List[str]:
+  parts = dutils.get_multi_sv_pair(uttr.detection)
+  if not parts or len(parts) != 2:
+    return []
+  if hoist_topics:
+    _hoist_topic(parts[0].svs)
+    _hoist_topic(parts[1].svs)
+  return parts[0].svs, parts[1].svs
+
+
+#
+# A topic may not often be the top-most result. In that case,
+# we look for a topic for up to TOPIC_RANK_LIMIT, and hoist to top
+# (This is the same limit NL interface uses for opening up topic).
```

Comment:
nit: if these are now unified, this comment becomes redundant?

Reply:
This comment is still relevant for now.